DM-41367: Make Butler server deployable

dhirving opened the pull request:
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

dhirving committed:
Add kubernetes health check endpoint to server

dhirving committed:
Provide a Butler configuration from the server

dhirving committed:
Add gafaelfawr authentication to the client

dhirving committed:
Make "/butler" URL prefix variable in client

For phalanx deployments, all butler servers will have the same hostname but live under different paths.

timj commented on tests/test_server.py:
```diff
@@ -91,6 +100,29 @@ def test_remote_butler(self):
+
+        # This is kind of a fragile test.  Butler's search logic does a lot of
+        # manipulations involving creating new ResourcePaths, and ResourcePath
+        # doesn't use httpx so we can't easily inject the TestClient in there.
```
If we decide that we want to do that we can consider changing ResourcePath to use `httpx` rather than `requests` (Obviously this is not half a day's work but might not be a huge amount of work).

dhirving commented:
The jenkins run just failed for this, I need to gate out test_authentication.py like the rest of the RemoteButler tests because it ends up implicitly loading httpx

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+    """Search the environment for a Rubin Science Platform access token.
+
+    The token may come from the following sources in this order:
+    1. The "BUTLER_RUBIN_ACCESS_TOKEN" environment variable.
```
I think there might need to be a blank line above here to make it render properly.

This reminds me that I have not told you how to build the documentation...

timj replied:
```suggestion
    1. The ``BUTLER_RUBIN_ACCESS_TOKEN`` environment variable.
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+
+    Returns
+    -------
+     access_token: `str | None`
```
```suggestion
     access_token: `str` or `None`
```
Sphinx needs the individual types to have back ticks.

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+    Parameters
+    ----------
+    server_url : `str`
+        URL of the Butler server that the caller intends to connect to
```
```suggestion
        URL of the Butler server that the caller intends to connect to.
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+       running outside the Rubin Science Platform.  This token will be sent
+       to EVERY server that we connect to, so be careful when connecting to
+       untrusted servers.
+    2. The "ACCESS_TOKEN" environment variable.
```
```suggestion
    2. The ``ACCESS_TOKEN`` environment variable.
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+       untrusted servers.
+    2. The "ACCESS_TOKEN" environment variable.
+       This environment variable is provided by the Rubin Science Platform
+       Jupyter notebooks.  It will only be returned if the given `server_url`
```
```suggestion
       Jupyter notebooks.  It will only be returned if the given ``server_url``
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+    Returns
+    -------
+     access_token: `str | None`
+        A Rubin Science Platform access token, or None if no token was
```
```suggestion
        A Rubin Science Platform access token, or `None` if no token was
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+from fnmatch import fnmatchcase
+from urllib.parse import urlparse
+
+SERVER_WHITELIST = ["*.lsst.cloud"]
```
Convention is to use leading underscore in these globals so that people are not tempted to ask for them outside this package.

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+    -------
+     access_token: `str | None`
+        A Rubin Science Platform access token, or None if no token was
+        configured in the environment
```
```suggestion
        configured in the environment.
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+
+    Returns
+    -------
+    header_map : dict[str, str]
```
```suggestion
    header_map : `dict` [`str`, `str`]
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+    Parameters
+    ----------
+    access_token : `str`
+        Rubin Science Platform access token
```
```suggestion
        Rubin Science Platform access token.
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+
+def get_authentication_headers(access_token: str) -> dict[str, str]:
+    """Return HTTP headers required for authenticating the user via Rubin
+    Science Platform's Gafaelfawr service
```
```suggestion
    Science Platform's Gafaelfawr service.
```

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,95 @@
+    Returns
+    -------
+    header_map : dict[str, str]
+        HTTP header names and values as a mapping from name -> value
```
```suggestion
        HTTP header names and values as a mapping from name to value.
```

timj commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -80,8 +94,15 @@ def __init__(
+            if access_token is not None:
+                auth_headers = get_authentication_headers(access_token)
+
+            headers = auth_headers | {"user-agent": f"{get_full_type_name(self)}/{__version__}"}
```
Cute. Alternatively:

```python
headers = {"user-agent": f"{get_full_type_name(self)}/{__version__}"}
headers.update(auth_headers)
```
Which might be clearer eventually if we are ever going to put more than `user-agent` in the second dict above.

timj commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -66,10 +68,22 @@ def __init__(
+        # configuration option refers to the directory containing the
+        # configuration file.  We allow this for the remote butler's URL so
+        # that the server doesn't have to know which hostname it is being
+        # accessed from
```
```suggestion
        # accessed from.
```

timj commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -66,10 +68,22 @@ def __init__(
         butler_config = ButlerConfig(config, searchPaths, without_datastore=True)
+        # There is a convention in Butler config files where <butlerRoot> in a
+        # configuration option refers to the directory containing the
+        # configuration file.  We allow this for the remote butler's URL so
```
We don't do the "two spaces after a period" thing in general 😄

timj commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -305,8 +326,7 @@ def _get_url(self, path: str, version: str = "v1") -> str:
         path : `str`
-            The relative path to the server endpoint. Should not include the
-            "/butler" prefix.
+            The relative path to the server endpoint
```
```suggestion
            The relative path to the server endpoint.
```

timj commented on requirements.txt:
```diff
@@ -21,3 +21,4 @@ pyarrow >= 0.16
 responses >= 0.12.0
 urllib3 >= 1.25.10
 fastapi
+safir >= 3.4.0
```
You also need to add this to pyproject.toml so that `pip install lsst-daf-butler` works. Ideally there would be an optional `server` definition in `pyproject.toml` so you can do `pip install lsst-daf-butler[server]`. If it's optional the test code will have to skip if safir is missing.

timj commented on tests/test_authentication.py:
```diff
@@ -0,0 +1,50 @@
+from contextlib import contextmanager
+from unittest.mock import patch
+
+from lsst.daf.butler.remote_butler._authentication import (
```
Given the `_` in the import path here the implication is that some of those would be lifted to `lsst.daf.butler.remote_butler` namespace. If the two package globals are private (they are not in `__all__`) then maybe the imports should be:

```python
from lsst.daf.butler.remote_butler import (
    get_authentication_headers,
    get_authentication_token_from_environment,
)
from lsst.daf.butler.remote_butler._authentication import (
    EXPLICIT_BUTLER_ACCESS_TOKEN_ENVIRONMENT_KEY,
    RSP_JUPYTER_ACCESS_TOKEN_ENVIRONMENT_KEY,
)
```
(and maybe even with leading underscores if they are really private). If they aren't private they will need docstrings.

timj commented on tests/test_authentication.py:
```diff
@@ -0,0 +1,50 @@
+    def test_jupyter_token_with_unsafe_server(self):
+        with _mock_env({RSP_JUPYTER_ACCESS_TOKEN_ENVIRONMENT_KEY: "token2"}):
+            token = get_authentication_token_from_environment("https://untrustedserver.com/butler")
+            assert token is None
```
In unittest land these should be:

```python
self.assertIsNone(token)
```
and 
```python
self.assertEqual(token, "token2")
```

timj reviewed (commented):
I've made a quick scan through the changes. I'll take another look when you've finished it. Lots of docstring tweaks and I need to show you how to build docs.

dhirving committed:
Change path prefix for butler server

For the purposes of the RSP, Butler is considered part of the "API
Aspect", so the path to it needs to start with /api/.  Handlers were
re-organized to be grouped under an APIRouter, since this prefix will
need to be configurable in the future.

dhirving commented:
Alright @timj , I think I got everything.  Other than your comments I changed it so that instead of `/butler` we're under `/api/butler` like Russ requested.  This will need to become variable per-deployment shortly, so to make that easier I aligned us with the safir template's pattern for declaring that prefix.  That required declaring a separate APIRouter instance for the external routes and splitting up the main server file into a few files.

timj commented on python/lsst/daf/butler/remote_butler/_authentication.py:
```diff
@@ -0,0 +1,98 @@
+       running outside the Rubin Science Platform.  This token will be sent
+       to EVERY server that we connect to, so be careful when connecting to
+       untrusted servers.
+    2. The ``ACCESS_TOKEN"`` environment variable.
```
```suggestion
    2. The ``ACCESS_TOKEN`` environment variable.
```

timj commented on python/lsst/daf/butler/remote_butler/server/_dependencies.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def factory_dependency() -> Factory:
+    return Factory(butler=_make_global_butler())
```
I wonder why codecov doesn't think this line is called...

dhirving replied:
Because it's not, we override the factory dependency with a test fixture in the unit tests

timj reviewed (approved):
Looks great.
